### ISSUE TITLE
Support on Windows (Experiment)

### ISSUE THREAD

**Issue by ghost** [lukeiwanski/tensorflow#202]
Due to some AMD cards are not supported by AMD-GPU-Pro, such as HD7900 series, I think that's important to work on a Windows support for AMD cards.

I am trying that...

**Comment by lukeiwanski** [comment]
Hi @kcauchy 

In principle.. there is ComputeCpp for Windows.

Essentially, we did not try windows in Eigen yet - that would be a first step to do.
Please let us know how did that go :)

**Comment by DuncanMcBain** [comment]
To elaborate, Eigen uses features that have been historically poorly supported by Visual Studio. However, I know that 2015 is a lot better in that regard so it might work now. Unfortunately ComputeCpp doesn't (yet) support Visual Studio 2017, but works well with 2015.

**Comment by ghost** [comment]
As my test results under VS 2017 on Windows 10, Eigen at least can be compiled.

```
-- Build type:        Release
-- Build site:        desktop-o107hvb
-- Build string:      unknownwin-19.11.25547.0-32bit
-- Enabled backends:  
-- Disabled backends: Cholmod,  UmfPack,  SuperLU,  PaStiX,  METIS,  SPQR,  Qt4 support,  Boost.Multiprecision,  GoogleHash,  Adolc,  MPFR C++,  fftw,  OpenGL, 
-- Default order:     Column-major
-- Maximal matrix/vector size: 320
-- SSE2:              Using architecture defaults
-- SSE3:              Using architecture defaults
-- SSSE3:             Using architecture defaults
-- SSE4.1:            Using architecture defaults
-- SSE4.2:            Using architecture defaults
-- AVX:               Using architecture defaults
-- FMA:               Using architecture defaults
-- AVX512:            Using architecture defaults
-- Altivec:           Using architecture defaults
-- VSX:               Using architecture defaults
-- ARM NEON:          Using architecture defaults
-- ARMv8 NEON:        Using architecture defaults
-- S390X ZVECTOR:     Using architecture defaults
-- C++11:             OFF
-- SYCL:              OFF
-- CUDA:              OFF
```

And the compilation:

```
========== Build: 767 succeeded, 0 failed, 0 up-to-date, 8 skipped ==========
```

It seems our target is **SYCL: ON** right? The 
compilation took over 1 hour on my 2690 v3 (1 processor only). I remember the entire tensorflow doesn't need over 1 hour on a shit machine under Ubuntu. I don't know whether it is the Eigen cached or Windows is *that* slow.

I will try on another 2695 v2 *2 workstation tomorrow with Visual Studio 2015. Hopefully it will be faster.

Here's the log:

* `cmake ..` [cmake.log](https://github.com/lukeiwanski/tensorflow/files/1649137/cmake.log)
* `devenv Eigen.sln /Build Release` [build.log](https://github.com/lukeiwanski/tensorflow/files/1649140/build.log)

**Comment by mirh** [comment]
> Due to some AMD cards are not supported by AMD-GPU-Pro, such as HD7900 series

Actually.. That should be no problem.

**Comment by ghost** [comment]
@mirh 

This is the official response from AMD support team: https://community.amd.com/thread/223669/

> Currently, ampdgpu-pro does not support OpenCL on SI cards including HD7900 series. AFAIK, there is no immediate plan for including SI support at this moment.

So, who told you that "no problem"?

**Comment by mirh** [comment]
I guess like they might have not included all the SI fixes in the closed driver, but I cannot see why forcing amdgpu open kernel driver plus [this](https://github.com/grmat/opencl-amd) shouldn't work. 

Worst case scenario anyway, you can always use fglrx.

**Comment by ghost** [comment]
@lukeiwanski @DuncanMcBain Eigen is also compilable under Visual Studio 2015.

I guess the target is "SYCL: ON" instead of OFF.

**Comment by DuncanMcBain** [comment]
Yes, which should be doable with the CMake bits currently in Eigen (I'm not able really to check right now, but IIRC should be something like -DEIGEN_TEST_SYCL=1 -DEIGEN_USE_CXX11=1).

**Comment by welnaseth** [comment]
I'm interested in windows support as well, and am willing to help if you need people to test/compile dependencies to verify everything needed for tensorflow to run with opencl works on windows.

**Comment by lukeiwanski** [comment]
@welnaseth the first step would be to test if Eigen is working on windows ( https://bitbucket.org/mehdi_goli/opencl/commits/branch/Eigen-Optimised-Tensor-Vector-Contraction ) 

Could you give it a spin?

**Comment by welnaseth** [comment]
I am travelling this weekend, but will be back near my computer Sunday night, I'll try then and report back :)

**Comment by welnaseth** [comment]
Alright, So first observation is that the windows support added in https://github.com/codeplaysoftware/computecpp-sdk/pull/51 to ComputeCPP did not make it into this eigen branch, i.e. the FindComputeCpp.cmake does not look for the windows library correctly. I've copy pasted FindComputeCpp.cmake from computecpp-sdk and will see if that works when compiling eigen on windows with SYCL support.

**Comment by welnaseth** [comment]
Ok, after getting the cmake to run, I've come across the first error. In Meta.h on lines 333-334:

`template<typename T>`
`Index size(const T& x) { return x.size(); }`

Produces error c2228 with the message "left of '.size' must have class/struct/union". I'm currently looking into it, but I'll admit c++ isn't my strongest language, so if you have any idea what needs to be done to correct this error, that would be great.

**Comment by lukeiwanski** [comment]
@welnaseth thanks for checking this out! I created a separate issue so we can track that particular problem separately.
